modules/cudastereo/src/DepthmapDenoiseWeightedHuber.cpp

```diff
@@ -0,0 +1,338 @@
+//! "A first-order primal-dual algorithm for convex problems with applications to imaging."
+
+
+#include "cuda/DepthmapDenoiseWeightedHuber.cuh"
+#include "opencv2/core/cuda_stream_accessor.hpp"
+#include "opencv2/core.hpp"
+
+
+using namespace std;
+using namespace cv;
+using namespace cv::cuda;
+namespace cv{
+    namespace cuda{
+        class DepthmapDenoiseWeightedHuberImpl : public DepthmapDenoiseWeightedHuber
+        {
+        public:
+            //CostVolume cv;//The cost volume we are attached to
+
+            DepthmapDenoiseWeightedHuberImpl(const GpuMat& visibleLightImage=GpuMat(),Stream cvStream=Stream::Null());
+            GpuMat operator()(InputArray ain, float epsilon, float theta);
+
+            GpuMat visibleLightImage;
+            //buffers
+            GpuMat _qx,_qy,_d,_a,_g,_g1,_gx,_gy;
+            GpuMat stableDepth;
+
+
+            //in case you want to do these explicitly
+            void allocate(int rows,int cols, InputArray gxin = GpuMat(), InputArray gyin = GpuMat());
+            void cacheGValues(InputArray visibleLightImage=GpuMat());
+
+        private:
+            int rows;
+            int cols;
+
+            void computeSigmas(float epsilon,float theta);
+
+            //internal parameter values
+            float sigma_d,sigma_q;
+
+            //flags
+            bool cachedG;
+            int alloced;
+            int dInited;
+
+            void setStream(Stream /*s*/){CV_Assert(!"Not Implemented");};
+            Stream getStream(){CV_Assert(!"Not Implemented"); return cvStream;};
+
+            void setAlpha(float /*alpha*/){CV_Assert(!"Not Implemented");};
+            float getAlpha(){CV_Assert(!"Not Implemented");return 0;};
+
+            void setBeta(float /*beta*/){CV_Assert(!"Not Implemented");};
+            float getBeta(){CV_Assert(!"Not Implemented");return 0;};
+
+        public:
+            Stream cvStream;
+        };
+    }
+}
+
+
+
+
+namespace cv{
+    namespace cuda{
+using namespace std;
+using namespace cv::cuda;
+
+DepthmapDenoiseWeightedHuberImpl::DepthmapDenoiseWeightedHuberImpl(const GpuMat& _visibleLightImage,
+                                                        Stream _cvStream) :
+                                                        visibleLightImage(_visibleLightImage),
+                                                        rows(_visibleLightImage.rows),
+                                                        cols(_visibleLightImage.cols),
+                                                        cvStream(_cvStream)
+{
+    alloced=0;
+    cachedG=0;
+    dInited=0;
+}
+
+Ptr<DepthmapDenoiseWeightedHuber>
+CV_EXPORTS createDepthmapDenoiseWeightedHuber(InputArray visibleLightImage, Stream cvStream){
+    return Ptr<DepthmapDenoiseWeightedHuber>(new DepthmapDenoiseWeightedHuberImpl(visibleLightImage.getGpuMat(),cvStream));
+}
+
+
+#define FLATALLOC(n,cv) n.create(1,cv.rows*cv.cols, CV_32FC1);n=n.reshape(0,cv.rows)
```

Comment:
There is `cv::cuda::createContinuous` function, which does the same and works bot for `Mat` and `GpuMat`.